### PR TITLE
Extend examples

### PR DESCRIPTION
**Pull request by diarmidmackenzie**
This PR covers:
- breaking out examples into 3 sets, covering Cannon local, Cannon worker & Ammo
- creating equivalent Ammo examples for each Cannon example
- as part of this, implementing Ammo equivalents of the utility compenents used in examples: force-pushable, grab, rain-of-entities.
- enhancement to Ammo driver to support configuration of restitution on bodies. 
- enhancement to Ammo driver to support spring constraints
- fixing VR grabbing for both Cannon and Ammo examples
- major rework of docs: break Cannon-specific info into separate CannonDriver.md and make top-level doc even-handed regarding Ammo & Cannon
- document warnings about Cannon Worker having lots of bugs at the moment
- include mention of new C-Frame physx repo, and presentation of physx as an alternative to aframe-physics-system, including some pros & cons

Work on this PR has also uncovered a few issues that it has not resolved - these have been raised separately for future investigation.  See issues #7, #8, #9, #10

If you want to see a live staged version of all the changes, ahead of this being merged, you can see that here:
https://diarmidmackenzie.github.io/aframe-physics-system/

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,309 +17,125 @@ Supports [CANNON.js](http://schteppe.github.io/cannon.js/) and [Ammo.js](https:/
...
-  <!-- Floor -->
-  <a-plane static-body></a-plane>
+### Cannon.js
 
-  <!-- Immovable box -->
-  <a-box static-body position="0 0.5 -5" width="3" height="1" depth="1"></a-box>
+- Can't handle collision with fast moving bodies, as it does not offer Continuous Collision Detection (CCD)
+- No support for collision filtering - all object pairs 
+- Restitution (bounciness) is a global property, rather than per-body
+- No stateful collision data - just a stream of events that starts when collision is happening, and stops when collision ends (compare Ammo, which offers distinct collide-start, collide-end events and collision state that can be queried at any time; no idea yet what PhysX offers here...)
```
Comment (wmurphyrd) [inline README.md:120]: So both collision filtering and stateful collision data for Ammo.js are features I'd added in [aframe-physics-extras](https://github.com/wmurphyrd/aframe-physics-extras) - seeing them here as limitations is a good confirmation of our previous discussion about moving those components into this repo

File: examples/README.md [hunk 2]
```diff
@@ -3,45 +3,17 @@
...
+| Demonstration of many features in a single example.          | [**OK**](cannon/sandbox.html)     | [**Broken**](cannon-worker/sandbox.html)           | [**OK**](ammo/sandbox.html)         |
+| Construct a [compound shape](../README.md#shape) and simulate collision with a ground plane. | [**OK**](cannon/compound.html)    | [**Broken**](cannon-worker/compound.html)          | [**OK**](ammo/compound.html) |
+| Demonstration of many constraints including cone twist, hinge, lock, point to point, slider (Ammo only) and distance (Cannon only) constraints. | [**OK**](cannon/constraints.html) | [**Broken**](cannon-worker/constraints.html) | [**OK**](ammo/constraints.html) |
+| Bounce simulation with [restitution (bounciness)](../README.md#system-configuration) of 1. | [**OK**](cannon/materials.html)   | [**Limited**](cannon-worker/materials.html)   | [**OK**](ammo/materials.html) |
+| Four vertical [springs](../README.md#spring) each between two boxes with an assortment of damping and stiffness values | [**OK**](cannon/springs.html)     | [**OK**](cannon-worker/springs.html)              | [**Limited**](ammo/spring.html) |
```
Comment (wmurphyrd) [inline examples/README.md:15]:
```suggestion
| Four vertical [springs](../README.md#spring) each between two boxes with an assortment of damping and stiffness values | [**OK**](cannon/spring.html)     | [**OK**](cannon-worker/spring.html)              | [**Limited**](ammo/spring.html) |
```

Both cannon spring links were 404

File: README.md (hunk 1 above)
Comment (diarmidmackenzie) [reply README.md:120]: You mean for Cannon js, I think?  We already have them for Ammo.

**Comment by kylebakerio** [comment]
throwing this link here, instead of just code comments, so that github hooks it in to that conversation for google and such: https://github.com/n5ro/aframe-physics-system/issues/171

File: README.md [hunk 3]
```diff
@@ -17,309 +17,125 @@ Supports [CANNON.js](http://schteppe.github.io/cannon.js/) and [Ammo.js](https:/
...
-</a-scene>
-```
 
-## Components
 
-### `dynamic-body` and `static-body`
+### Ammo.js
 
-The `dynamic-body` and `static-body` components may be added to any `<a-entity/>` that contains a mesh. Generally, each scene will have at least one `static-body` for the ground, and one or more `dynamic-body` instances that the player can interact with.
+- No support for off-center attachment of constraints to bodies (integration issue?)
```
Comment (kylebakerio) [inline README.md:126]: e.g., hinging a door? (if so, maybe add that line?)

File: README.md [hunk 4]
```diff
@@ -17,309 +17,125 @@ Supports [CANNON.js](http://schteppe.github.io/cannon.js/) and [Ammo.js](https:/
...
 
-- **dynamic-body**: A freely-moving object. Dynamic bodies have mass, collide with other objects, bounce or slow during collisions, and fall if gravity is enabled.
-- **static-body**: A fixed-position or animated object. Other objects may collide with static bodies, but static bodies themselves are unaffected by gravity and collisions.
+- No support for slider constraint (integration issue?)
```
Comment (kylebakerio) [inline README.md:128]: maybe a quick comment explaining what a slider is

File: README.md [hunk 5]
```diff
@@ -17,309 +17,125 @@ Supports [CANNON.js](http://schteppe.github.io/cannon.js/) and [Ammo.js](https:/
...
-  <a-entity camera look-controls></a-entity>
+This list is probably incomplete, so if you find an additional significant limitation, please add it to this list.
 
-  <!-- Floor -->
-  <a-plane static-body></a-plane>
+### Cannon.js
 
-  <!-- Immovable box -->
-  <a-box static-body position="0 0.5 -5" width="3" height="1" depth="1"></a-box>
+- Can't handle collision with fast moving bodies, as it does not offer Continuous Collision Detection (CCD)
```
Comment (kylebakerio) [inline README.md:117]: I wrote my own CCD (no physics, but just for collision detection) recently, would be cool to figure out how to add this. Probably I should look upstream, first have been meaning to package the CCD as-is as a standalone package.

**Review by kylebakerio — COMMENTED** [review]
review just on initial read through, haven't tried examples yet.

File: CannonDriver.md [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+CDN builds for aframe-physics-system@v$npm_package_version:
+
+- [aframe-physics-system.js](https://cdn.jsdelivr.net/gh/c-frame/aframe-physics-system@v4.1.0/dist/aframe-physics-system.js) *(development)*
+- [aframe-physics-system.min.js](https://cdn.jsdelivr.net/gh/c-frame/aframe-physics-system@v4.1.0/dist/aframe-physics-system.min.js) *(production)*
+
+### npm
+
+```
+npm install --save aframe-physics-system
```
Comment (kylebakerio) [inline CannonDriver.md:25]: this is out of date, not our version, right?

File: README.md (hunk 1 above)
Comment (diarmidmackenzie) [reply README.md:120]: I'll add a reference to aframe-physics-extras

File: dist/aframe-physics-system.js [hunk 7]
```diff
@@ -15972,7 +15989,8 @@ let AmmoBody = {
     disableCollision: { default: false },
     collisionFilterGroup: { default: 1 }, //32-bit mask,
     collisionFilterMask: { default: 1 }, //32-bit mask
-    scaleAutoUpdate: { default: true }
+    scaleAutoUpdate: { default: true },
+    restitution: {default: 0} // does not support updates
```
[the remark this answers is not in the document]
Comment (diarmidmackenzie) [reply dist/aframe-physics-system.js:15993]: Agreed

File: examples/ammo/materials.html [hunk 8]
```diff
@@ -0,0 +1,48 @@
+<html>
+  <head>
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta http-equiv="content-type" content="text/html; charset=utf-8">
+    <meta name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no,user-scalable=no,maximum-scale=1">
+    <title>Examples • Materials AMMO</title>
+    <script src="https://aframe.io/releases/1.3.0/aframe.min.js"></script>
+    <script src="https://unpkg.com/aframe-environment-component@1.3.1/dist/aframe-environment-component.min.js"></script>
+    <script src="https://cdn.jsdelivr.net/gh/donmccurdy/aframe-extras@v6.1.1/dist/components/sphere-collider.min.js"></script>
```
[the remark this answers is not in the document]
Comment (diarmidmackenzie) [reply examples/ammo/materials.html:9]: Yes.  But I've realized we don't need this at all... 

I updated grab component to work based on physics collisions, so we no longer need the "hit" event therefore no longer need any sphere-colliders.  I think I can just remove across the board with no ill effects, so I'll do that.

File: examples/components/rain-of-entities.js [hunk 9]
```diff
@@ -41,12 +41,8 @@ AFRAME.registerComponent('rain-of-entities', {
 
     // Recycling is important, kids.
```
[the remark this answers is not in the document]
Comment (diarmidmackenzie) [reply examples/components/rain-of-entities.js:42]: not mine :-)

File: README.md [hunk 10]
```diff
@@ -17,309 +17,125 @@ Supports [CANNON.js](http://schteppe.github.io/cannon.js/) and [Ammo.js](https:/
...
+
+- a-frame-physics system with Cannon driver
+- a-frame-physics system with Ammo driver
+- [physx](https://github.com/c-frame/physx) (which uses Nvidia physX as its driver).  Note that there is no current plan to integrate physX into aframe-physics-system, but it may be a better choice for some projects.
+
+Since each driver has a slightly different component interface and schema, it will require some significant updates to your code to switch from one driver to another, so it's worth taking some time up-front to consider which driver is most likely to suit your needs.
+
+At a high level:
+
+- The Cannon driver is the easiest to use, but as a native JavaScript solution, it has the worst performance
```
[the remark this answers is not in the document]
Comment (diarmidmackenzie) [reply README.md:35]: Have run perf tests for ammo vs physX.  Never ran perf tests for cannon.  The code that outputs timing data isn't in this fork yet - I've raised #13 to cover this.

File: README.md (hunk 3 above)
Comment (diarmidmackenzie) [reply README.md]: HInges are fine - restriction only applies to spring constraints (now clarified).

File: README.md (hunk 4 above)
Comment (diarmidmackenzie) [reply README.md]: done.  Like a bead on an abacus.

File: README.md (hunk 5 above)
Comment (kylebakerio) [reply README.md:117]: of course, just a comment.

File: CannonDriver.md (hunk 6 above)
Comment (kylebakerio) [reply CannonDriver.md:25]: reminder on this one

File: README.md (hunk 4 above)
Comment (kylebakerio) [reply README.md]: Not sure if there was a mistake, but I don't see this update.

**Comment by kylebakerio** [comment]
Love the examples, great work.

I get no reticle on the contraints for ammo page, the way I do on constraints for cannon. Are you aware of that and chose not to do it, or is it an error or oversight? Not a reason to deny the pull request, but nice to have if you were intending it.

**Comment by kylebakerio** [comment]
Would also be nice if the examples were separated under 'cannon' and 'ammo' categories.

I also feel like the 'attached to cone' in the initial examples for both is pretty awkward... just makes a bad first impression and looks broken. Not saying you should rework that (happy to have it if you are willing), but we should probably do that. Not sure what we should take away from that, but seems like just flipping the cone upside down and having cube on the flat side would be cleaner?

Am I alone in thinking this?

File: examples/ammo/materials.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (diarmidmackenzie) [reply examples/ammo/materials.html]: Now done this (and re-tested a couple of examples in VR - going to assume the others are fine too as it's the same code).

**Comment by diarmidmackenzie** [comment]
> Would also be nice if the examples were separated under 'cannon' and 'ammo' categories.

Did you look at this page?

https://diarmidmackenzie.github.io/aframe-physics-system/examples/

I've realized that this page...
https://diarmidmackenzie.github.io/aframe-physics-system/examples/

... has a link to the non-updated examples here:
https://c-frame.github.io/aframe-physics-system/examples/

So maybe that caused some confusion?

**Comment by diarmidmackenzie** [comment]
> I get no reticle on the contraints for ammo page

https://diarmidmackenzie.github.io/aframe-physics-system/examples/ammo/constraints.html

(see prev. comment, maybe you ended up one one of the non-updated pages due to incorrect absolute link)

File: README.md (hunk 4 above)
Comment (diarmidmackenzie) [reply README.md]: Maybe I forgot to push.

https://github.com/c-frame/aframe-physics-system/pull/11/commits/c712774436ed3deeadd3d634e3c1b36a421b0bf0

File: CannonDriver.md (hunk 6 above)
Comment (diarmidmackenzie) [reply CannonDriver.md]: npm install in general looks pretty flaky / in need of testing.  But I think that's separate from this PR so I'm going to leave for now.

**Comment by diarmidmackenzie** [comment]
> I also feel like the 'attached to cone' in the initial examples for both is pretty awkward...

> Yes, agreed. I'll try to do something to improve this.

Just moving the cone 1m to the left so the 2 objects are near to each other but don't intersect or z-fight makes this look a lot better IMO.  I still don't love the example, but I'd rather work on new, better examples, than invest in this one any further...

**Comment by kylebakerio** [comment]
I'm sure the incorrect links were the issue. 

Going to the ammo example now, I get a reticle, but I get broken examples.

image

...then, after pasting this image and being here, and flipping back to the tab in question, suddenly the example are now working.

image

Interesting. My internet is very slow here today, so that is probably the cause, figured I'll post anyways since it's interesting that that's possible.

**Comment by kylebakerio** [comment]
Is there anything left pending before this is ready for merge?

**Comment by kylebakerio** [comment]
image

**Pull request merged**